File: Python/543.py
# How does this problem work?
    # Recursive solution where leafs are the base case
    # Set a nonlocal variable res to keep track of the max diameter
    # We want to find the node with the highest combined left/right depth
    # At any given node, check the combined depth of the left/right subtree
    # After this, return 1 + max(left, rigth) to find the current node's
        # maximum depth and continue the algorithm a level higher
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(root):
            nonlocal res
            if not root: return 0
            left = dfs(root.left)
            right = dfs(root.right)
            res = max(res, left + right)
            return 1 + max(left, right)

        dfs(root)
        return res

File: C++/543.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int result = 0;
        dfs(root, result);
        return result;
    }
private:
    int dfs(TreeNode* root, int& result){
        if(root == NULL) return 0;
        int left = dfs(root -> left, result);
        int right = dfs(root -> right, result);
        result = max(result, left + right);
        return 1 + max(left, right);
    }
};
/*
- Idea - keep a result variable
    - for each node, recursively calculate the depth of each subtree
    - update result if the sum of the subtrees is greater than the current maximum
*/